gap/SnAnBB.gi

```diff
@@ -18,100 +18,917 @@
 ##  The code is based upon the algorithm presented in the paper [BLGN+03].
 ##
 ##
-# DeclareInfoClass( "InfoRecSnAn" );
-# SetInfoLevel( InfoRecSnAn, 0 );
-#
-# # The following are used for comparisons:
-#
-# # FIXME: Get rid of these globals.
-# # However, it seems to me this whole file is dead code anyway..
-# RecSnAnIsOne := IsOne;
-# RecSnAnEq := EQ;
-#
-# ## is <r,s> in S_n
-#
-# SatisfiesSnPresentation := function( n, r, s )
-#
-#     local j, t;
+BindGlobal("SatisfiesSnPresentation",
+function(ri, n, r, s)
+    local j, t;
+    if not(isone(ri)((r * s)^(n-1))) then
+        return false;
+    fi;
+    j := 2;
+    t := r;
+    while j <= n/2 do
+        t := t * r;
+        if not isone(ri)(Comm(s,t)^2) then
+            return false;
+        fi;
+        j := j + 1;
+    od;
+    return true;
+end);
+
+BindGlobal("SatisfiesAnPresentation",
+function(ri, s, t, n)
+    local j, r, ti;
+
+    if not(isone(ri)(s^(n-2))) or not(isone(ri)(t^3)) then
+        return false;
+    fi;
+
+    if n mod 2 <> 0 then
+        # we already know s^(n-2) = t^3 = 1
+        if not(isone(ri)((s * t)^n)) then
+            return false;
+        fi;
+        j := 1;
+        r := s^0;
+        while j <= (n-3)/2  do
+            r := r * s;
+            if not(isone(ri)((t *(t^r))^2)) then
+                return false;
+            fi;
+            j := j + 1;
+        od;
```

Comment:
Optimization suggestion (avoids recomputing `(n-3)/2`, and the GAP kernel has a special fast case to compute `for i in RANGE` loops; also saves one multiplication `r * s` per loop iteration; you do need to add `tr` to the list of local variables, though)
```suggestion
        tr := t;
        for j in [1..(n-3)/2] do
            tr := tr ^ s; # equal to t^(s^j)
            if not isone(ri)((t * tr)^2) then
                return false;
            fi;
        od;
```